File: pocketbase/pb_hooks/utils/notifications.js
/// <reference path="../../pb_data/types.d.ts" />

function usersByPermission(app, permission) {
    try {
        return app.findRecordsByFilter(
            'users',
            `role.permissions.name ?= "${permission}"`,
            '',
            200,
            0,
        )
    } catch (err) {
        app.logger().error(
            'notifications: failed to resolve users by permission',
            'permission',
            permission,
            'error',
            String(err),
        )
        return []
    }
}

function push(app, { users, type, params, url }) {
    if (!users || !users.length) return 0

    let collection
    try {
        collection = app.findCollectionByNameOrId('notifications')
    } catch (err) {
        app.logger().error(
            'notifications: collection missing',
            'error',
            String(err),
        )
        return 0
    }

    let written = 0
    for (const user of users) {
        try {
            const record = new Record(collection)
            record.set('user', user.id)
            record.set('type', type)
            record.set('params', params || {})
            record.set('url', url || '')
            record.set('read', false)
            app.save(record)
            written++
        } catch (err) {
            app.logger().error(
                'notifications: failed to queue',
                'type',
                type,
                'user',
                user.id,
                'error',
                String(err),
            )
        }
    }

    return written
}

module.exports = { usersByPermission, push }

File: pocketbase/hooks/notifications.go
package hooks

import (
	"github.com/pocketbase/dbx"
	"github.com/pocketbase/pocketbase/core"
)

const (
	readNotificationRetentionDays = 30
	maxNotificationRetentionDays  = 90
)

type notification struct {
	Users  []*core.Record
	Type   string
	Params map[string]any
	URL    string
}

func registerNotifications(app core.App) {
	app.Cron().MustAdd("notificationRetention", "23 3 * * *", func() {
		removed, err := pruneRows(
			app,
			"DELETE FROM notifications WHERE (`read` = TRUE AND created < {:readCutoff}) OR created < {:maxCutoff}",
			dbx.Params{
				"readCutoff": cutoff(days(readNotificationRetentionDays)),
				"maxCutoff":  cutoff(days(maxNotificationRetentionDays)),
			},
		)
		if err != nil {
			app.Logger().Error("notifications: prune failed", "error", err)
			return
		}
		app.Logger().Info("notifications: pruned expired entries", "rows", removed)
	})
}

func usersByPermission(app core.App, permission string) []*core.Record {
	users, err := app.FindRecordsByFilter(
		"users",
		"role.permissions.name ?= {:permission}",
		"",
		200,
		0,
		dbx.Params{"permission": permission},
	)
	if err != nil {
		app.Logger().Error("notifications: failed to resolve users by permission", "permission", permission, "error", err)
		return nil
	}
	return users
}

func pushNotification(app core.App, message notification) int {
	if len(message.Users) == 0 {
		return 0
	}

	collection, err := app.FindCollectionByNameOrId("notifications")
	if err != nil {
		app.Logger().Error("notifications: collection missing", "error", err)
		return 0
	}

	params := message.Params
	if params == nil {
		params = map[string]any{}
	}

	written := 0
	for _, user := range message.Users {
		record := core.NewRecord(collection)
		record.Set("user", user.Id)
		record.Set("type", message.Type)
		record.Set("params", params)
		record.Set("url", message.URL)
		record.Set("read", false)
		if err := app.Save(record); err != nil {
			app.Logger().Error("notifications: failed to queue", "type", message.Type, "user", user.Id, "error", err)
			continue
		}
		written++
	}
	return written
}
